codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/endpoints/DefaultEndpointProviderGenerator.kt

```diff
@@ -170,9 +185,13 @@ class DefaultEndpointProviderGenerator(
         withConditions(rule.conditions) {
             writer.withBlock("return #T(", ")", RuntimeTypes.SmithyClient.Endpoints.Endpoint) {
                 writeInline("#T.parse(", RuntimeTypes.Core.Net.Url.Url)
+                renderingEndpointUrl = true
                 renderExpression(rule.endpoint.url)
+                renderingEndpointUrl = false
```

Comment:
**Style**: Doing this kind of mutable state tracking via class-level fields is difficult to follow and error-prone. The visitor pattern allows using a return type to get information from within the tree but it looks like we currently use `ExpressionVisitor<Unit>` and skip returning anything.

My suggestion would be to refactor this to keep a context result in the visitor:

```kotlin
data class EndpointInfo(val params: MutableSet<String>) {
    companion object {
        val Empty = EndpointInfo(params = mutableSetOf())
    }

    operator fun plus(other: EndpointInfo) = EndpointInfo(
        params = this.params + other.params,
    )
}

class ExpressionGenerator(…): ExpressionVisitor<EndpointInfo>, … {
    …

    override fun visitRef(reference: Reference): EndpointInfo {
        val referenceName = reference.name.defaultName()
        val isParamReference = isParamRef(reference)

        if (isParamReference) {
            writer.writeInline("params.")
        }
        writer.writeInline(referenceName)

        return if (isParamReference) {
            EndpointInfo(params = mutableSetOf(referenceName))
        } else {
            EndpointInfo.Empty
        }
    }

    …
}
```

Then you can modify the expression renderer to return/use the information:

```
fun interface ExpressionRenderer {
    fun renderExpression(expr: Expression): EndpointInfo
}

class DefaultEndpointProviderGenerator(…): ExpressionRenderer {
    …

    override fun renderExpression(expr: Expression): EndpointInfo = expr.accept(expressionGenerator)

    private fun renderEndpointRule(rule: EndpointRule) {
        withConditions(rule.conditions) {
            writer.withBlock("return #T(", ")", RuntimeTypes.SmithyClient.Endpoints.Endpoint) {
                writeInline("#T.parse(", RuntimeTypes.Core.Net.Url.Url)
                val endpointInfo = renderExpression(rule.endpoint.url)
                write("),")

                val hasAccountIdBasedEndpoint = "accountId" in endpointInfo.params
                val hasServiceEndpointOverride = "endpoint" in endpointInfo.params
                val needAdditionalEndpointProperties = hasAccountIdBasedEndpoint || hasServiceEndpointOverride

                if (rule.endpoint.headers.isNotEmpty()) { … }

                if (rule.endpoint.properties.isNotEmpty() || needAdditionalEndpointProperties) {
                    withBlock("attributes = #T {", "},", RuntimeTypes.Core.Collections.attributesOf) {
                        rule.endpoint.properties.entries.forEach { (k, v) -> … }

                        if (hasAccountIdBasedEndpoint) {
                            writer.write("#T to true", RuntimeTypes.Core.BusinessMetrics.accountIdBasedEndPoint)
                        }
                        if (hasServiceEndpointOverride) {
                            writer.write("#T to true", RuntimeTypes.Core.BusinessMetrics.serviceEndpointOverride)
                        }
                    }
                }
            }
        }
    }

    …
}


```

Reply:
@0marperez Were you able to take a look at this feedback, do you think it's possible to refactor? I also agree that the boolean toggling stands out and seems error-prone for future implementations

Reply:
I ran into some difficulties while trying to implement this and decided to leave it as is since the comment is marked as `style`. I think it's possible so I'll give it another go